xbmc/network/httprequesthandler/python/HTTPSessionManager.cpp

```diff
@@ -0,0 +1,261 @@
+/*
+ *      Copyright (C) 2015 Team XBMC
+ *      http://xbmc.org
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <time.h>
+
+#include "HTTPSessionManager.h"
+#include "interfaces/legacy/mod_python/HttpRequest.h"
+#include "interfaces/legacy/mod_python/Session.h"
+#include "threads/SingleLock.h"
+#include "utils/log.h"
+#include "utils/md5.h"
+#include "utils/StringUtils.h"
+#include "utils/SystemInfo.h"
+
+#define SESSION_SUFFIX  "sid"
+
+using namespace XBMCAddon::xbmcmod_python;
+
+CHTTPSessionManager::CHTTPSessionManager()
+{ }
+
+CHTTPSessionManager::~CHTTPSessionManager()
+{
+  CSingleLock lock(m_critSection);
+  for (std::map<std::string, Session*>::const_iterator session = m_sessions.begin(); session != m_sessions.end(); ++session)
+    delete session->second;
+  m_sessions.clear();
+}
+
+CHTTPSessionManager& CHTTPSessionManager::Get()
+{
+  static CHTTPSessionManager m_instance;
+  return m_instance;
+}
+
+bool CHTTPSessionManager::IsValidSessionId(const std::string& sessionId)
+{
+  return sessionId.size() == 32 &&
+         sessionId.find_first_not_of("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ") == std::string::npos;
+}
+
+std::string CHTTPSessionManager::GenerateSessionId(XBMCAddon::xbmcmod_python::HttpRequest* request)
+{
+  if (request == NULL || request->m_request == NULL || request->m_request->connection == NULL)
+    return "";
+
+  // get the ip address of the requester
+  const MHD_ConnectionInfo *ipInfo = MHD_get_connection_info(request->m_request->connection, MHD_CONNECTION_INFO_CLIENT_ADDRESS);
+
+  // generate a new session id
+  std::string sessionId;
+  int count = 0;
+  CSingleLock lock(m_critSection);
+  do
+  {
+    std::string temp = StringUtils::Format("%d%d%d%s", time(NULL) * (10000 + count), rand(), rand(), inet_ntoa(ipInfo->client_addr->sin_addr));
```

Comment:
‘struct sockaddr’ has no member named ‘sin_addr’. A quick grep showed a lot of casting to `struct sockaddr_in` (I have no idea what any of this is)